### ISSUE TITLE
Compiler hangs when pattern matching an enum variant that does not exist.

### ISSUE THREAD

**Issue by SimonImbrogno** [rust-lang/rust#91165]
I tried this code:

```rust
enum Foo {
    Bar,
    Baz,
}

fn main() {
    let x = Foo::Bar;

    match x {
        Foo::BadVariant => { /* Big Yikes */ },
        _ => {}
    }
}
```

**I expected to see this happen:** 
I expected an error indicating that the variant `BadVariant` does not exist in `Foo`

**Instead, this happened:** 
The compiler hangs.

**Note:**
I also tried using an `if let` instead of a `match` but got the same result. 
I'm surprised I've not heard of anyone encountering this before, so I fear its related to some specific quirk of my system. 😬 

Worth mentioning that _only_ this scenario seems to cause the hang. The compilation is finishing and printing diagnostics with no issues when other errors are introduced instead.

### Meta

I tried nightly(`1.58.0`), stable(`1.56.1`) and `1.55.0`. `cargo clean` + fresh build. Same issue with all three.

`rustc --version --verbose`:
```
rustc 1.56.1 (59eed8a2a 2021-11-01)
binary: rustc
commit-hash: 59eed8a2aac0230a8b53e89d4e99d55912ba6b35
commit-date: 2021-11-01
host: x86_64-pc-windows-msvc
release: 1.56.1
LLVM version: 13.0.0
```
```
rustc 1.58.0-nightly (936f2600b 2021-11-22)
binary: rustc
commit-hash: 936f2600b6c903b04387f74ed5cbce88bb06d243
commit-date: 2021-11-22
host: x86_64-pc-windows-msvc
release: 1.58.0-nightly
LLVM version: 13.0.0
```

<details><summary>Backtrace</summary>
<p>

The compiler hangs.
No backtrace is ever printed.

</p>
</details>

**Comment by wesleywiser** [comment]
That is very strange! I can't repro on either `x86_64-pc-windows-msvc` or `x86_64-unknown-linux-gnu` for any of stable, beta or nightly. All toolchains report:

```
error[E0599]: no variant or associated item named `BadVariant` found for enum `Foo` in the current scope
  --> src/main.rs:10:14
   |
1  | enum Foo {
   | -------- variant or associated item `BadVariant` not found here
...
10 |         Foo::BadVariant => { /* Big Yikes */ },
   |              ^^^^^^^^^^ variant or associated item not found in `Foo`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `hang` due to previous error
```

Perhaps there is some quirk of whatever terminal/shell you're using?

**Comment by SimonImbrogno** [comment]
> Perhaps there is some quirk of whatever terminal/shell you're using?

Ha, some strange quirk indeed. Nobody n the discord could seem to repro it either. I don't think it's my shell (I'm using `cmder`), because the background `cargo check` that rust-analyzer runs was hanging as well. (I made sure to kill all rust related processes before trying to repro in case something there was interfering.)

However, since yesterday, I've rebooted my system and tried compiling again. It _seems_ to have resolved itself. I cannot say why... The compiler was using basically no system resourced when it hung... so it wasn't in a busy loop or chewing through memory or anything... my only thought is that specific code path makes some blocking OS call that was just... not returning?  

I don't know why on earth pattern matches specifically were causing it though. The compiler seemed to run as expected in _all other instances_. 

🤷 The bits were not cooperating yesterday...  

Shall I close the issue then? I'm not sure if there's a tag for "weird one-off nonsense that nobody can repro" 😁

**Comment by wesleywiser** [comment]
I think it makes sense to close at this point since you can no longer reproduce this either. If you (or anyone else) experiences this again, we can always re-open. 

For future reference, the next step is probably to attach a debugger and collect backtraces to see what is blocked.
